### ISSUE TITLE
LLDB expr can't handle C++ template template parameters.

### ISSUE THREAD

**Issue by awulkiew** [llvm/llvm-project#61725]
OS: Ubuntu 22.04.2 LTS
LLDB version: 14.0.0

The code:
```
#include <vector>

template <typename T, template<typename, typename> class TT>
struct B : TT<T, std::allocator<T>> {};
struct D : B<int, std::vector> {};

int main()
{
    D d;

    return 0;
}
```
LLDB:
```
(lldb) expr (B<int, std::vector>*)(&d)
error: expression failed to parse:
error: <user expression 0>:1:14: use of class template 'std::vector' requires template arguments
(B<int, std::vector>*)(&d)
             ^
```

**Comment by awulkiew** [comment]
@dwblaikie AFAIU this is not what you mean but just to be clear, this can be reproduced with any template, not only `std::vector`.

**Comment by dwblaikie** [comment]
Yeah, I see in the example now - yeah, there's a type with that name, it's reasonable to expect to be able to refer to that existing name even given dwarfs limitations here. Sorry for the confusion.

**Comment by awulkiew** [comment]
Yes, it's an existing name, i.e. base class. Without it the name `std::vector` cannot be found at all.

This example is a simplified version of Boost.Geometry container-based models, e.g. 
[boost::geometry::multi_polygon](https://github.com/boostorg/geometry/blob/boost-1.81.0/include/boost/geometry/geometries/multi_polygon.hpp#L55).

I wanted to cast a pointer to object of this type to pointer to its base class to get a container that is used internally in order to fix support for LLDB in this vscode extension: https://github.com/awulkiew/graphical-debugging-vscode but due to this issue I was forced to do it in a different way.

**Comment by Michael137** [comment]
Ah looks like in `TypeSystemClang.cpp` `CreateTemplateParameterList` we only ever create either `NonTypeTemplateParmDecl` (for `TemplateArgument::Kind::Integeral` only) or `TemplateTypeParmDecl`. Whereas we do create `TemplateArgument`s of kind `Template` when parsing DWARF. So I think that's where the mismatch comes from

**Comment by Michael137** [comment]
The DWARF for `Vector` looks like this:
```
DW_TAG_structure_type
  DW_AT_name      ("B<Vector>")
  DW_TAG_GNU_template_template_param
    DW_AT_name    ("TT")
    DW_AT_GNU_template_name       ("Vector")
```

LLDB does parse this and creates a `clang::TemplateTemplateParmDecl` via `TypeSystemClang::CreateTemplateTemplateParmDecl`. When we then create a `ClassTemplateDecl` we call `CreateTemplateParameterList` which iterates over the `TemplateArgument`s we just parsed from DWARF. `CreateTemplateParameterList` only ever creates `NonTypeTemplateParmDecl` (for `TemplateArgument::Kind::Integral` only) and `TemplateTypeParmDecl` (even for `TemplateArgument::Kind::Template`). This creates the discrepancy down the line.

However, to be able to create a `TemplateTemplateParmDecl` we need a `TemplateParameterList` that describes `Vector`s parameters. Just from DWARF we don't have that info. We only have the template name.
